### PR TITLE
Standardize GetHashCode implementation

### PR DESCRIPTION
**Pull request by utpilla**
## Changes
- Use `HashCode.Add` and `HashCode.ToHashCode` methods to generate the HashCode on `netstandard2.1` and `net6.0` and greater
- Add zero for values that could be null
- Add missing unchecked blocks

### CODE REVIEW

**Review by Kielek — APPROVED** [review]

File: src/OpenTelemetry/Trace/SamplingResult.cs [hunk 1]
```diff
@@ -117,11 +117,23 @@ public override bool Equals(object? obj)
...
+            hashCode.Add(this.TraceStateString?.GetHashCode() ?? 0);
+
+            var hash = hashCode.ToHashCode();
+#else
+            var hash = 17;
+            unchecked
+            {
+                hash = (31 * hash) + this.Decision.GetHashCode();
+                hash = (31 * hash) + this.Attributes.GetHashCode();
+                hash = this.TraceStateString == null ? hash : (hash * 31) + this.TraceStateString.GetHashCode();
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Trace/SamplingResult.cs:133]: I don't think it is a big deal, but if you want to be consistent with the others...

```csharp
hash = (31 * hash) + this.TraceStateString?.GetHashCode() ?? 0;
```

File: src/OpenTelemetry/Metrics/MetricStreamIdentity.cs [hunk 2]
```diff
@@ -37,9 +37,30 @@ public MetricStreamIdentity(Instrument instrument, MetricStreamConfiguration met
...
+#if NETSTANDARD2_1 || NET6_0_OR_GREATER
+            HashCode hashCode = default;
+            hashCode.Add(this.InstrumentType.GetHashCode());
+            hashCode.Add(this.MeterName.GetHashCode());
+            hashCode.Add(this.MeterVersion.GetHashCode());
+            hashCode.Add(this.InstrumentName.GetHashCode());
+            hashCode.Add(this.HistogramRecordMinMax.GetHashCode());
+            hashCode.Add(this.Unit.GetHashCode());
+            hashCode.Add(this.Description.GetHashCode());
+            hashCode.Add(this.ViewId?.GetHashCode() ?? 0);
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/MetricStreamIdentity.cs:49]: I was just looking at the [source to HashCode.Add](https://github.com/dotnet/runtime/blob/dcda1e08cb8f421877388be2d685c009a31b1845/src/libraries/System.Private.CoreLib/src/System/HashCode.cs#L304) it seems to do the coalescing automatically so you could probably simplify this stuff to just pass the raw value eg:

```csharp
hashCode.Add(this.ViewId);
```

File: src/OpenTelemetry/Metrics/MetricStreamIdentity.cs [hunk 3]
```diff
@@ -37,9 +37,30 @@ public MetricStreamIdentity(Instrument instrument, MetricStreamConfiguration met
...
+            hashCode.Add(this.TagKeys != null ? StringArrayComparer.GetHashCode(this.TagKeys) : 0);
```
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/MetricStreamIdentity.cs:50]: There is an overload which accepts `IEqualityComparer<T>` you might be able to use here:

```csharp
hashCode.Add(this.TagKeys, StringArrayComparer);
```

**Review by CodeBlanch — APPROVED** [review]
Some nits but LGTM

File: src/OpenTelemetry.Api/Baggage.cs [hunk 4]
```diff
@@ -345,17 +345,17 @@ public override int GetHashCode()
         {
             var baggage = this.baggage ?? EmptyBaggage;
 
-            unchecked
+            var hash = 17;
+            foreach (var item in baggage)
```
Comment (alanwest) [inline src/OpenTelemetry.Api/Baggage.cs:349]: Could be another PR, but in a few places we compute the hash in the constructor since often the state is immutable. I think Baggage is one such case we could do this.

**Review by alanwest — APPROVED** [review]

File: src/OpenTelemetry.Api/Baggage.cs (hunk 4 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Api/Baggage.cs:349]: We might want to be selective about where we do that. Because, GetHashCode is probably never called on Baggage. Unless someone was using it as a key somewhere, which seems unlikely? We really only want to do that computation ahead of time where we know for sure it will be used as key. Otherwise it would be just wasted cycles 😄

Comment (alanwest) [reply src/OpenTelemetry.Api/Baggage.cs:349]: Good point, makes sense.

File: src/OpenTelemetry/Metrics/MetricStreamIdentity.cs (hunk 3 above)
Comment (utpilla) [reply src/OpenTelemetry/Metrics/MetricStreamIdentity.cs]: Updated.

File: src/OpenTelemetry/Trace/SamplingResult.cs (hunk 1 above)
Comment (utpilla) [reply src/OpenTelemetry/Trace/SamplingResult.cs]: Update this.

I have also made this change for `MetricStreamIdentity` class as well:

```csharp
hash = (hash * 31) + this.Unit?.GetHashCode() ?? 0;
hash = (hash * 31) + this.Description?.GetHashCode() ?? 0;
hash = (hash * 31) + this.ViewId ?? 0;
hash = (hash * 31) + (this.TagKeys != null ? StringArrayComparer.GetHashCode(this.TagKeys) : 0);
```
